mbs-s0015-20240504/example_usage.py

```diff
@@ -0,0 +1,63 @@
+import numpy as np
+from pixell import enmap, curvedsky, enplot
+from os.path import join
+
+# Note, this script it set up to be run on NERSC.
+
+# first get some basic info like the path and the filename template
+fdir = '/global/cfs/cdirs/sobs/v4_sims/mbs/mbs_s0015_20240504/sims'
+fbase_template = 'so_lat_mbs_mss0002_{model}_{bands}_lmax5400_4way_set{split_num}_noise_sim_map{sim_num:04}.fits'
+
+# For models, pick from 'fdw_lf', 'fdw_mf' or 'fdw_uhf'. The corresponding entries for
+# the bands are: 'lf_f030_lf_f040', 'mf_f090_mf_f150' and uhf_f230_uhf_f290.
+
+# Let's load the sim for e.g. f150, split 2, sim 123
+model = 'fdw_mf'
+bands = 'mf_f090_mf_f150'
+band_idx = 1                # f090 is 0, f150 is 1
+split_num = 2
+sim_num = 123               # has 4 digits (leading 0's), but padding done for us in template
+
+sim = enmap.read_map(
+    join(fdir, fbase_template.format(
+        model=model,
+        bands=bands,
+        split_num=split_num,
+        sim_num=sim_num)
+        ),
+    sel=np.s_[band_idx, 0]  # just load f150, remove split axis
+    )
+print(sim.geometry, sim.dtype)
+
+assert sim.shape == (3, 2640, 10800)
+assert np.allclose(sim.wcs.wcs.cdelt, [-0.03333333, 0.03333333])
+assert np.allclose(sim.wcs.wcs.crval, [180, 0])
+assert np.allclose(sim.wcs.wcs.crpix, [5400.62, 1890.50])
+assert sim.dtype == np.float32
+
+# The sim only contains noise power to the Nyquist frequency of the pixelization (for 2 arcmin pixels, `lmax=5400`).
+# But what if we wanted to work with objects in map-space defined at the full resolution of the data, for example a
+# sky mask defined at 0.5 arcmin resolution? We need to project the `sim` to our desired geometry.
+
+# Because the `sim` is bandlimited, we can do this losslessly with a spherical harmonic transform:
+
+# First we need the geometry of the pixelization.
+# We can get this from one of the time-domain simulations
+shape, wcs = enmap.read_map_geometry(
+    '/global/cfs/cdirs/sobs/sims/mss-0002/RC1.r01/sobs_RC1.r01_LAT_mission_f150_4way_split1_noise_map_car.fits')
+
+# Allocate an empty map to hold the sim at full resolution
+sim_fullres = enmap.empty((*sim.shape[:-2], *shape[-2:]), wcs, sim.dtype)
+
+# Project sim to alm, then from alm to map
+
+# High res vesion:
+#curvedsky.alm2map(curvedsky.map2alm(sim, lmax=5400), sim_fullres)
+
+# Low res version for quick testing:
+curvedsky.alm2map(curvedsky.map2alm(sim, lmax=300), sim_fullres)
```

Comment:
pixell '0.17.3' gives:

```
((3, 2640, 10800), car:{cdelt:[-0.03333,0.03333],crval:[180,0],crpix:[5400.62,1890.50]}) float32
---------------------------------------------------------------------------
ShapeError                                Traceback (most recent call last)
File /global/u2/z/zonca/p/software/map_based_simulations/mbs-s0015-20240504/example_usage.py:58
     50 sim_fullres = enmap.empty((*sim.shape[:-2], *shape[-2:]), wcs, sim.dtype)
     52 # Project sim to alm, then from alm to map
     53 
     54 # High res vesion:
     55 #curvedsky.alm2map(curvedsky.map2alm(sim, lmax=5400), sim_fullres)
     56 
     57 # Low res version for quick testing:
---> 58 curvedsky.alm2map(curvedsky.map2alm(sim, lmax=300), sim_fullres)
     60 # Finally, we can plot the sim:
     61 for pol in range(3):

File /global/common/software/cmb/zonca/conda/pycmb/lib/python3.10/site-packages/pixell/curvedsky.py:146, in map2alm(map, alm, ainfo, lmax, spin, direct, copy, oversample, method, tweak, rtol, atol, alm2map_adjoint)
    125 """Spherical harmonics analysis of the enmap map[...,ny,nx] into the spherical harmonics
    126 coefficients alm[...,nalm]. The (approximate) inverse of alm2map. To support partial
    127 sky coverage and arbitrary projections, an intermediate map will be constructed
   (...)
    143 
    144 The alm2map_adjoint argument is used internally to implement the alm2map_adjoint function."""
    145 if method == "cyl" or method == "auto":
--> 146 	alm = map2alm_cyl(map, alm, ainfo=ainfo, lmax=lmax, spin=spin, direct=direct,
    147 			copy=copy, tweak=tweak, rtol=rtol, atol=atol, alm2map_adjoint=alm2map_adjoint)
    148 elif method == "pos":
    149 	raise NotImplementedError("map2alm for noncylindrical layouts not implemented")

File /global/common/software/cmb/zonca/conda/pycmb/lib/python3.10/site-packages/pixell/curvedsky.py:292, in map2alm_cyl(map, alm, ainfo, lmax, spin, direct, copy, tweak, rtol, atol, alm2map_adjoint)
    289 	tmap[tslice] = map[mslice]
    290 # We don't have ring weights for general cylindrical projections.
    291 # See if our pixelization matches one with known weights.
--> 292 minfo = get_minfo(tmap.shape, tmap.wcs, quad=True, tweak=tweak, rtol=rtol, atol=atol)
    293 return map2alm_raw(tmap, alm, minfo, ainfo, spin=spin, copy=copy, alm2map_adjoint=alm2map_adjoint)

File /global/common/software/cmb/zonca/conda/pycmb/lib/python3.10/site-packages/pixell/curvedsky.py:553, in get_minfo(shape, wcs, quad, tweak, rtol, atol)
    539 """Get a map info to be used in a spherical harmonics transform for the
    540 given geometry (shape, wcs).
    541 quad: Specifies whether we need quadrature weights or not. These are only
   (...)
    550 rtol and atol: Controls the maximum relative and absolute tolerance allowed
    551  for tweaking, in units of pixels. Defaults to None which implies 1 pixel."""
    552 if tweak: return match_predefined_minfo(shape, wcs, rtol=rtol, atol=atol)
--> 553 if quad:  return match_predefined_minfo(shape, wcs, rtol=1e-6, atol=1e-6)
    554 else:     return geo2minfo(shape, wcs)

File /global/common/software/cmb/zonca/conda/pycmb/lib/python3.10/site-packages/pixell/curvedsky.py:621, in match_predefined_minfo(shape, wcs, rtol, atol)
    619 roff /= pixheight
    620 if not roff < rtol: raise ShapeError("Could not find a map_info with predefined quadrature weights matching input map (rel offset %e >= %e). Pass tweak=True to map2alm to allow it to use slightly shifted pixel positions to match a geometry for which a predefined quadrature exists. This matches the old default. The resulted slightly shifted alms can be projected back onto the original pixels by passing tweak=True to alm2map." % (aoff, atol))
--> 621 if not aoff < atol: raise ShapeError("Could not find a map_info with predefined quadrature weights matching input map (abs offset %e >= %e). Pass tweak=True to map2alm to allow it to use slightly shifted pixel positions to match a geometry for which a predefined quadrature exists. This matches the old default. The resulted slightly shifted alms can be projected back onto the original pixels by passing tweak=True to alm2map." % (aoff, atol))
    622 minfo = minfos2[best]
    623 # Modify the minfo to restrict it to only the rows contained in the geometry

ShapeError: Could not find a map_info with predefined quadrature weights matching input map (abs offset 5.000000e-01 >= 1.000000e-06). Pass tweak=True to map2alm to allow it to use slightly shifted pixel positions to match a geometry for which a predefined quadrature exists. This matches the old default. The resulted slightly shifted alms can be projected back onto the original pixels by passing tweak=True to alm2map.
```

Reply:
if you need a specific version of `pixell`, please put an assert for that as well

Reply:
Added an assert statement to the script.